drivers/flash/soc_flash_mcux.c

```diff
@@ -205,35 +205,55 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
 	addr = offset + priv->pflash_block_base;
 
 #ifdef CONFIG_CHECK_BEFORE_READING
+	/*
+	 * Ensure the area is readable, since a direct access may cause faults
+	 * on erased or otherwise unreadable pages. Emulate erased pages,
+	 * return other errors.
+	 */
   #ifdef CONFIG_SOC_LPC55S36
-	/* Validates the given address range is loaded in the flash hiding region. */
-	rc = FLASH_IsFlashAreaReadable(&priv->config, addr, len);
-	if (rc != kStatus_FLASH_Success) {
-		rc = -EIO;
-	} else {
-		/* Check whether the flash is erased ("len" and "addr" must be word-aligned). */
-		rc = FLASH_VerifyErase(&priv->config, ((addr + 0x3) & ~0x3),  ((len + 0x3) & ~0x3));
+	/* On LPC55S36, use a HAL function to safely copy from Flash. */
+	rc = FLASH_Read(&priv->config, addr, data, len);
+	switch (rc) {
+	case kStatus_FLASH_Success:
+		rc = 0;
+		break;
+	case kStatus_FLASH_EccError:
+		/* Check id the ECC issue is due to the Flash being erased
+		 * ("addr" and "len" must be word-aligned for this call).
+		 */
+		rc = FLASH_VerifyErase(&priv->config,
```

Comment:
Question: wouldn't it be better to just ifdef the `is_area_readable` internally to use this method of verification, whether memory is readable, for `CONFIG_SOC_LPC55S36`, and then follow it with the memcopy or memset depending on result? Instead of reading first and then checking why it failed, and have streamlined code fall all devices supported?

As far as I understand the description for `FLASH_VerifyErase` in fsl_iap.h, it is able to tell whether there is ECC problem in range (btw NXP, the function no longer takes margin as parameter, yet you still mention it in description).

Additional note: IMHO making read check configurable makes no sense, because there is always a chance that ECC error will topple device instead of returning with error, unless of course you disable ECC,